nav2_map_server/src/vo_server/vector_object_server.cpp

```diff
@@ -0,0 +1,558 @@
+: nav2_util::LifecycleNode("vector_object_server", "", options), process_map_(false)
+{}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  // Transform buffer and listener initialization
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    this->get_node_base_interface(),
+    this->get_node_timers_interface());
+  tf_buffer_->setCreateTimerInterface(timer_interface);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  // Obtaining ROS parameters
+  if (!obtainParams()) {
+    return nav2_util::CallbackReturn::FAILURE;
+  }
+
+  map_pub_ = create_publisher<nav_msgs::msg::OccupancyGrid>(
+    map_topic_,
+    rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable());
+
+  add_shapes_service_ = create_service<nav2_msgs::srv::AddShapes>(
+    "~/add_shapes",
+    std::bind(&VectorObjectServer::addShapesCallback, this, _1, _2, _3));
+
+  get_shapes_service_ = create_service<nav2_msgs::srv::GetShapes>(
+    "~/get_shapes",
+    std::bind(&VectorObjectServer::getShapesCallback, this, _1, _2, _3));
+
+  remove_shapes_service_ = create_service<nav2_msgs::srv::RemoveShapes>(
+    "~/remove_shapes",
+    std::bind(&VectorObjectServer::removeShapesCallback, this, _1, _2, _3));
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  map_pub_->on_activate();
+
+  // Trigger map to be published
+  process_map_ = true;
+  switchMapUpdate();
+
+  // Creating bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // Destroying bond connection
+  destroyBond();
+
+  if (map_timer_) {
+    map_timer_->cancel();
+    map_timer_.reset();
+  }
+  process_map_ = false;
+
+  map_pub_->on_deactivate();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  add_shapes_service_.reset();
+  get_shapes_service_.reset();
+  remove_shapes_service_.reset();
+
+  map_pub_.reset();
+  map_.reset();
+
+  shapes_.clear();
+
+  tf_listener_.reset();
+  tf_buffer_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+VectorObjectServer::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+bool VectorObjectServer::obtainParams()
+{
+  // Main ROS-parameters
```

Comment:
Usually we do a `auto node = shared_from_this()` so that the `node` is used in all the calls below (its more natural looking)